src/main/resources/migration/changelog-v5.6.0.xml

```diff
@@ -1398,4 +1398,124 @@
             <column name="NAME"/>
         </createIndex>
     </changeSet>
+
+    <changeSet id="v5.6.0-20" author="nscuro">
+        <!--
+          Disable deferred constraint checking during this transaction.
+          Without this, we would not be able to add primary keys after
+          having deleted records from a table before.
+        -->
+        <sql>SET CONSTRAINTS ALL IMMEDIATE</sql>
```

Comment:
I did a final test run with duplicate records across all tag tables, and the migration ended up failing with:

```
cannot ALTER TABLE "NOTIFICATIONRULE_TAGS" because it has pending trigger events
```

This happened when the migration tried to add primary keys. It turns out that, if a table has deferred foreign key constraints, you cannot modify rows and alter the table schema in the same transaction: https://github.com/friek/til/blob/master/postgres/alter-table-pending-triggers.md

The solution is to disable deferred constraints during the transaction, such that foreign keys are checked immediately and cannot block table altering.